### PR TITLE
Test: Mock Input columns for operator tests

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 1]
```diff
@@ -298,6 +298,8 @@ class DAGContext
     }
 
     bool isTest() const { return is_test; }
+    void setColumnsForTest(ColumnsWithTypeAndName columns_for_test_) { columns_for_test = columns_for_test_; }
+    ColumnsWithTypeAndName columnsForTest() { return columns_for_test; }
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGContext.h:302]: How do we support `select * from a join b;`
such as two `exchange_receiver`

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 2]
```diff
@@ -159,13 +159,34 @@ AnalysisResult analyzeExpressions(
...
+        if (context.getDAGContext()->columnsForTest().empty())
+        {
+            auto mock_table_scan_stream = std::make_shared<MockTableScanBlockInputStream>(columns_with_type_and_name, context.getSettingsRef().max_block_size);
+            pipeline.streams.emplace_back(mock_table_scan_stream);
+        }
+        else
+        {
+            auto mock_table_scan_stream = std::make_shared<MockTableScanBlockInputStream>(context.getDAGContext()->columnsForTest(), context.getSettingsRef().max_block_size);
+            pipeline.streams.emplace_back(mock_table_scan_stream);
+        }
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:189]: how about
```
auto columns_with_type_and_name = context.getDAGContext()->columnsForTest().empty() ? xxx : context.getDAGContext()->columnsForTest();

for (size_t i = 0; i < max_streams; ++i)
{
    auto stream = xxxxxxxx;
    pipeline.streams.emplace_back(mock_table_scan_stream);
}
```

File: dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp [hunk 3]
```diff
@@ -272,8 +272,8 @@ void CreatingSetsBlockInputStream::createOne(SubqueryForSet & subquery)
                 if (subquery.table)
                     msg.fmtAppend("Table with {} rows. ", head_rows);
 
-                msg.fmtAppend("In {.3f} sec. ", watch.elapsedSeconds());
-                msg.fmtAppend("using {} threads.", subquery.join ? subquery.join->getBuildConcurrency() : 1);
+                msg.fmtAppend("In {0:.3f} sec. ", watch.elapsedSeconds());
+                msg.append("using 1 threads.");
```
Comment (SeaRise) [inline dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp:276]: why not `subquery.join ? subquery.join->getBuildConcurrency() : 1)`

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 4]
```diff
@@ -298,6 +298,8 @@ class DAGContext
     }
 
     bool isTest() const { return is_test; }
+    void setColumnsForTest(std::unordered_map<String, ColumnsWithTypeAndName> & columns_for_test_map_) { columns_for_test_map = columns_for_test_map_; }
+    ColumnsWithTypeAndName columnsForTest(String exector_id) { return columns_for_test_map[exector_id]; }
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGContext.h:302]: throw Exception when `columns_for_test_map.find(executor_id) == columns_for_test_map.end()`?

File: dbms/src/TestUtils/InterpreterTestUtils.cpp [hunk 5]
```diff
@@ -65,6 +65,70 @@ void InterpreterTest::executeInterpreter(const String & expected_string, const s
...
+    for (const auto & block : blocks)
+    {
+        if (!sample_block)
+        {
+            sample_block = block;
+            for (const auto & column : block.getColumnsWithTypeAndName())
+            {
+                actual_cols.push_back(column.type->createColumn());
+            }
+        }
```
Comment (SeaRise) [inline dbms/src/TestUtils/InterpreterTestUtils.cpp:84]:
```suggestion
    Block sample_block = blocks.back();
    std::vector<MutableColumnPtr> actual_cols;
    for (const auto & column : sample_block.getColumnsWithTypeAndName())
    {
        actual_cols.push_back(column.type->createColumn());
    }
    for (const auto & block : blocks)
    {
```

**Comment by ywqzzy** [comment]
Need to consider a way to test interpreter without user defined columns, I will fix it up today

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 6]
```diff
@@ -156,16 +156,60 @@ AnalysisResult analyzeExpressions(
 }
 } // namespace
 
-// for tests, we need to mock tableScan blockInputStream as the source stream.
-void DAGQueryBlockInterpreter::handleMockTableScan(const TiDBTableScan & table_scan, DAGPipeline & pipeline)
+void DAGQueryBlockInterpreter::mockSourceStream(String executor_id, DAGPipeline & pipeline, const String & source_type)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:159]: why not 
```
template<typename SourceType>
void DAGQueryBlockInterpreter::mockSourceStream(String executor_id, DAGPipeline & pipeline)
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 7]
```diff
@@ -266,7 +310,7 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
         stream->setExtraInfo(
             fmt::format("join build, build_side_root_executor_id = {}", dagContext().getJoinExecuteInfoMap()[query_block.source_name].build_side_root_executor_id));
     });
-    executeUnion(build_pipeline, max_streams, log, /*ignore_block=*/true, "for join");
+    executeUnion(build_pipeline, max_streams, log, !dagContext().isTest(), "for join");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:313]:
```suggestion
    executeUnion(build_pipeline, max_streams, log, /*ignore_block=*/!dagContext().isTest(), "for join");
```

and add some comments

File: dbms/src/TestUtils/InterpreterTestUtils.h [hunk 8]
```diff
@@ -52,6 +52,16 @@ class InterpreterTest : public ::testing::Test
 
     void executeInterpreter(const String & expected_string, const std::shared_ptr<tipb::DAGRequest> & request, size_t concurrency);
 
+    void executeStreams(
+        const std::shared_ptr<tipb::DAGRequest> & request,
+        std::unordered_map<String, ColumnsWithTypeAndName> & source_columns_map,
+        const ColumnsWithTypeAndName & expect_columns,
+        size_t concurrency);
```
Comment (SeaRise) [inline dbms/src/TestUtils/InterpreterTestUtils.h:59]:
```suggestion
        size_t concurrency = 1);
```

File: dbms/src/TestUtils/InterpreterTestUtils.h [hunk 9]
```diff
@@ -52,6 +52,16 @@ class InterpreterTest : public ::testing::Test
...
+    void executeStreams(
+        const std::shared_ptr<tipb::DAGRequest> & request,
+        const ColumnsWithTypeAndName & expect_columns,
+        size_t concurrency);
```
Comment (SeaRise) [inline dbms/src/TestUtils/InterpreterTestUtils.h:63]:
```suggestion
        size_t concurrency = 1);
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 10]
```diff
@@ -266,7 +308,7 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
         stream->setExtraInfo(
             fmt::format("join build, build_side_root_executor_id = {}", dagContext().getJoinExecuteInfoMap()[query_block.source_name].build_side_root_executor_id));
     });
-    executeUnion(build_pipeline, max_streams, log, /*ignore_block=*/true, "for join");
+    executeUnion(build_pipeline, max_streams, log, /*ignore_block=*/!dagContext().isTest(), "for join");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:311]: add some comments like `for test, we should xxx output xxx`?

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 11]
```diff
@@ -83,7 +83,7 @@ BlockIO InterpreterDAG::execute()
     DAGPipeline pipeline;
     pipeline.streams = streams;
     /// add union to run in parallel if needed
-    if (dagContext().isMPPTask())
+    if (dagContext().isMPPTask() && !dagContext().isTest())
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:86]: refactor like
```
if (dagContext().isTest()
    executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/true, "for test");
else if (dagContext().isMPPTask())
    executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/true, "for mpp");
else
    executeUnion(pipeline, max_streams, dagContext().log, false, "for non mpp");
```
?

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+                        toNullableVec<String>("s", {"banana", "banana"}),
+                        toNullableVec<String>("join_c", {"apple", "banana"})},
+                       5);
+
+        executeStreams(request,
+                       {toNullableVec<String>("s", {"banana", "banana"}),
+                        toNullableVec<String>("join_c", {"apple", "banana"}),
+                        toNullableVec<String>("s", {"banana", "banana"}),
+                        toNullableVec<String>("join_c", {"apple", "banana"})},
+                       1);
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_executor.cpp:213]:
```suggestion
                        toNullableVec<String>("join_c", {"apple", "banana"})});
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 13]
```diff
@@ -156,16 +156,58 @@ AnalysisResult analyzeExpressions(
 }
 } // namespace
 
-// for tests, we need to mock tableScan blockInputStream as the source stream.
-void DAGQueryBlockInterpreter::handleMockTableScan(const TiDBTableScan & table_scan, DAGPipeline & pipeline)
+template <typename SourceType>
+void DAGQueryBlockInterpreter::mockSourceStream(String executor_id, DAGPipeline & pipeline)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:160]: can we move the related code of `mockSourceStream` to a separate file?

File: dbms/src/Flash/Coprocessor/MockSourceStream.h [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    ColumnsWithTypeAndName columns_with_type_and_name;
+    NamesAndTypes names_and_types;
+    size_t rows = 0;
+    std::vector<std::shared_ptr<SourceType>> mock_source_streams;
+    columns_with_type_and_name = context.getDAGContext()->columnsForTest(executor_id);
+    for (const auto & col : columns_with_type_and_name)
+    {
+        if (rows == 0)
+            rows = col.column->size();
+        RUNTIME_ASSERT(rows == col.column->size(), log, "each column has same size");
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/MockSourceStream.h:36]:
```suggestion
        RUNTIME_ASSERT(rows == col.column->size(), log, "each column must has same size");
```

File: dbms/src/TestUtils/mockExecutor.h [hunk 15]
```diff
@@ -122,18 +123,33 @@ class MockDAGRequestContext
...
+
     DAGRequestBuilder scan(String db_name, String table_name);
     DAGRequestBuilder receive(String exchange_name);
 
 private:
     size_t index;
     std::unordered_map<String, MockColumnInfos> mock_tables;
     std::unordered_map<String, MockColumnInfos> exchange_schemas;
+    std::unordered_map<String, ColumnsWithTypeAndName> mock_table_columns;
+    std::unordered_map<String, ColumnsWithTypeAndName> mock_exchange_columns;
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/TestUtils/mockExecutor.h:151]: o... if receiver.name == table.name...

File: dbms/src/Flash/Coprocessor/MockSourceStream.h [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+            columns_for_stream.push_back(
+                ColumnWithTypeAndName(
+                    column_with_type_and_name.column->cut(start, row_for_current_stream),
+                    column_with_type_and_name.type,
+                    column_with_type_and_name.name));
+        }
+        start += row_for_current_stream;
+        mock_source_streams.emplace_back(std::make_shared<SourceType>(columns_for_stream, context.getSettingsRef().max_block_size));
+    }
+    return {names_and_types, mock_source_streams};
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/MockSourceStream.h:57]: my mistake...

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/MockSourceStream.h:57]: assert(start == rows)？

File: dbms/src/Flash/tests/gtest_executor.cpp [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+// limitations under the License.
+
+#include <TestUtils/ExecutorTestUtils.h>
+#include <TestUtils/mockExecutor.h>
+
+namespace DB
+{
+namespace tests
+{
+class ExecutorTestDemo : public DB::tests::ExecutorTest
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_executor.cpp:22]: still demo?

File: dbms/src/TestUtils/ExecutorTestUtils.cpp [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        }
+    }
+
+    ColumnsWithTypeAndName actual_columns;
+    for (size_t i = 0; i < actual_cols.size(); ++i)
+        actual_columns.push_back({std::move(actual_cols[i]), sample_block.getColumnsWithTypeAndName()[i].type, sample_block.getColumnsWithTypeAndName()[i].name, sample_block.getColumnsWithTypeAndName()[i].column_id});
+    return Block(actual_columns);
+}
+
+void readBlock(BlockInputStreamPtr stream, const ColumnsWithTypeAndName & expect_columns)
```
Comment (SeaRise) [inline dbms/src/TestUtils/ExecutorTestUtils.cpp:96]: move to `namespace {}`？

**Review by SeaRise — APPROVED** [review]
excellent job!
LGTM

File: dbms/src/Flash/tests/gtest_executor.cpp (hunk 17 above)
Comment (ywqzzy) [reply dbms/src/Flash/tests/gtest_executor.cpp:22]: > still demo?

I will come up with a better name for it.

File: dbms/src/TestUtils/mockExecutor.cpp [hunk 19]
```diff
@@ -274,52 +276,112 @@ DAGRequestBuilder & DAGRequestBuilder::buildAggregation(ASTPtr agg_funcs, ASTPtr
...
+{
+    addExchangeRelationSchema(name, columnInfos);
+    addExchangeReceiverColumnData(name, columns);
+}
+
 DAGRequestBuilder MockDAGRequestContext::scan(String db_name, String table_name)
 {
-    return DAGRequestBuilder(index).mockTable({db_name, table_name}, mock_tables[db_name + "." + table_name]);
+    auto builder = DAGRequestBuilder(index).mockTable({db_name, table_name}, mock_tables[db_name + "." + table_name]);
+    if (mock_table_columns.find(db_name + "." + table_name) != mock_table_columns.end())
```
Comment (mengxin9014) [inline dbms/src/TestUtils/mockExecutor.cpp:370]: It will do nothing if not find  the table name?

File: dbms/src/TestUtils/mockExecutor.cpp [hunk 20]
```diff
@@ -274,52 +276,112 @@ DAGRequestBuilder & DAGRequestBuilder::buildAggregation(ASTPtr agg_funcs, ASTPtr
...
+        executor_id_columns_map[builder.getRoot()->name] = mock_table_columns[db_name + "." + table_name];
+    }
+    return builder;
 }
 
 DAGRequestBuilder MockDAGRequestContext::receive(String exchange_name)
 {
     auto builder = DAGRequestBuilder(index).exchangeReceiver(exchange_schemas[exchange_name]);
     receiver_source_task_ids_map[builder.getRoot()->name] = {};
+    if (mock_exchange_columns.find(exchange_name) != mock_exchange_columns.end())
```
Comment (mengxin9014) [inline dbms/src/TestUtils/mockExecutor.cpp:381]: ditto

File: dbms/src/TestUtils/mockExecutor.cpp (hunk 19 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/mockExecutor.cpp:370]: Yes. In this case,  user must put input columns in executeStreams in order to run the query.
Maybe I need to comment it out.

File: dbms/src/TestUtils/ExecutorTestUtils.cpp [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+    context = MockDAGRequestContext(TiFlashTestEnv::getContext());
+    dag_context_ptr->log = Logger::get("executorTest");
+}
+
+void ExecutorTest::SetUpTestCase()
+{
+    try
+    {
+        DB::registerFunctions();
+        DB::registerAggregateFunctions();
```
Comment (mengxin9014) [inline dbms/src/TestUtils/ExecutorTestUtils.cpp:40]: Add registerWindowFunctions?

File: dbms/src/TestUtils/mockExecutor.cpp (hunk 19 above)
Comment (mengxin9014) [reply dbms/src/TestUtils/mockExecutor.cpp:370]: ok.

File: dbms/src/TestUtils/ExecutorTestUtils.cpp [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+        actual_blocks.push_back(block);
+    }
+    stream->readSuffix();
+    Block actual_block = mergeBlocks(actual_blocks);
+    ASSERT_BLOCK_EQ(except_block, actual_block);
+}
+
+void ExecutorTest::executeStreams(const std::shared_ptr<tipb::DAGRequest> & request, std::unordered_map<String, ColumnsWithTypeAndName> & source_columns_map, const ColumnsWithTypeAndName & expect_columns, size_t concurrency)
+{
+    DAGContext dag_context(*request, "interpreter_test", concurrency);
```
Comment (mengxin9014) [inline dbms/src/TestUtils/ExecutorTestUtils.cpp:115]: You rename InterpreterTestUtils.cpp to ExecutorTestUtil.cpp , If here should be changed accordingly.

File: dbms/src/TestUtils/ExecutorTestUtils.cpp (hunk 21 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/ExecutorTestUtils.cpp:40]: The code here may have potential bug in unit test if we run function related test and MockExecutor test simultaneously(https://github.com/pingcap/tiflash/issues/5081), I wonder if I can fix it in another pr.

**Review by mengxin9014 — APPROVED** [review]

**Pull request merged**